File: crates/cxx-qt-lib/src/gui/qregion.cpp
// clang-format off
// SPDX-FileCopyrightText: 2024 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// clang-format on
// SPDX-FileContributor: Laurent Montel <laurent.montel@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

#include "cxx-qt-lib/qregion.h"

#include "../assertion_utils.h"

// https://code.qt.io/cgit/qt/qtbase.git/tree/src/gui/painting/qregion.h?h=v5.15.6-lts-lgpl#n178
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/gui/painting/qregion.h?h=v6.2.4#n161
constexpr static ::std::array<::std::size_t, 1> arr{ sizeof(::std::size_t) };
assert_alignment_and_size(QRegion, alignof(::std::size_t), arr, arr.size());

static_assert(!::std::is_trivially_copy_assignable<QRegion>::value);
static_assert(!::std::is_trivially_copy_constructible<QRegion>::value);

static_assert(!::std::is_trivially_destructible<QRegion>::value);

static_assert(QTypeInfo<QRegion>::isRelocatable);

File: tests/qt_types_standalone/rust/src/qregion.rs
// SPDX-FileCopyrightText: 2024 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// SPDX-FileContributor: Laurent Montel <laurent.montel@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

use cxx_qt_lib::QRegion;

#[cxx::bridge]
mod qregion_cxx {
    unsafe extern "C++" {
        include!("cxx-qt-lib/qregion.h");

        type QRegion = cxx_qt_lib::QRegion;
    }

    extern "Rust" {
        fn construct_qregion() -> QRegion;
        fn clone_qregion(p: &QRegion) -> QRegion;
    }
}

fn construct_qregion() -> QRegion {
    QRegion::default()
}

fn clone_qregion(p: &QRegion) -> QRegion {
    p.clone()
}
